asgiproxy/config.py

```diff
@@ -36,6 +37,21 @@ def process_client_headers(self, *, scope: Scope, headers: Headers) -> Headerlik
         """
         return headers
 
+    def get_client_protocols(self, *, scope: Scope, headers: Headers) -> Iterable[str]:
+        """
+        Get client subprotocol list so it can be passed upstream.
+        """
+        return scope.get("subprotocols", [])
+
+    def get_params(self, *, scope: Scope, headers: Headers) -> dict:
+        """
+        Get client query stream so it can be passed upstream.
+        """
+        qs = scope.get("query_string", b"")
+        if not qs:
+            return {}
+        return dict(QueryParams(qs))
```

Comment:
I don't think we should be assuming the query string is in a format that can be parsed by `QueryParams`, much less return a `dict`; for instance, `?foo;bar;blap` is a valid query string, as is `?foo=1&foo=2&foo=3`.

Reply:
Thanks for looking at this! Would it be appropriate if we updated `get_upstream_websocket_options` to use `get_upstream_url_with_query` instead then? I can try that now.